src/webgpu/api/validation/encoding/encoder_open_state.spec.ts

```
@@ -0,0 +1,383 @@
+          }
+          break;
+        case 'dispatchWorkgroups':
+          {
+            const computePipeline = t.createNoOpComputePipeline();
+            computePass.setPipeline(computePipeline);
+            computePass.dispatchWorkgroups(0);
+          }
+          break;
+        case 'dispatchWorkgroupsIndirect':
+          {
+            const computePipeline = t.device.createComputePipeline({
+              layout: 'auto',
+              compute: {
+                module: t.device.createShaderModule({
+                  code: `
+                    @group(0) @binding(0) var outImage : texture_storage_2d<rgba8unorm, write>;
+
+                    @compute @workgroup_size(1) fn main() {
+                      textureStore(outImage, vec2<i32>(0, 0), vec4<f32>(1.0, 0.0, 0.0, 1.0));
+                    }`,
+                }),
+                entryPoint: 'main',
+              },
+            });
+
+            const kDispatchIndirectParametersSize = 12;
+            const indirectBuffer = t.device.createBuffer({
+              size: kDispatchIndirectParametersSize,
+              usage: GPUBufferUsage.INDIRECT,
+            });
+            t.trackForCleanup(indirectBuffer);
+
+            const outputTexture = t.device.createTexture({
+              format: 'rgba8unorm',
+              size: [1, 1, 1],
+              usage:
+                GPUTextureUsage.COPY_SRC |
+                GPUTextureUsage.RENDER_ATTACHMENT |
+                GPUTextureUsage.STORAGE_BINDING,
+            });
+            t.trackForCleanup(outputTexture);
+
+            const bindGroup = t.device.createBindGroup({
+              layout: computePipeline.getBindGroupLayout(0),
+              entries: [
+                {
+                  binding: 0,
+                  resource: outputTexture.createView(),
+                },
+              ],
+            });
+
+            computePass.setBindGroup(0, bindGroup);
+            computePass.setPipeline(computePipeline);
```

These commands would need to be encoded before ending the pass, so that we're testing validation inside dispatchWorkgroups only. However, I don't think we actually need to set the preconditions: if those fail, they'll just make the pass/encoder invalid. Here we are testing whether the encoding command *itself* generates an error. We don't even need to end the pass or finish the encoder at the end at all.

Same for other commands that have preconditions, like endOcclusionQuery.

---

Thank you for pointing it out. I remove unnecessary preconditions in the tests.